### ISSUE TITLE
3.9.4: FTBFS@i686 with "-DENABLE_TESTS=YES"

### ISSUE THREAD

**Issue by onlyjob** [lizardfs/lizardfs#197]
On i686 LizardFS FTBFS when built with "-DENABLE_TESTS=YES":

~~~~
Building CXX object src/common/CMakeFiles/mfscommon_tests.dir/serialization_macros_unittest.cc.o 
cd /tmp/buildd/lizardfs-2.5.4.2/build/src/common && /usr/lib/ccache/c++   -DLIZARDFS_HAVE_CONFIG_H -DTHROW_INSTEAD_OF_ABORT -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -D_FORTIFY_SOURCE=2  -pipe -std=c++0x -pthread -Wall -Wextra -fwrapv -pedantic -Wno-gnu -mcrc32 -O3 -DNDEBUG -O3 -DNDEBUG -g -I/tmp/buildd/lizardfs-2.5.4.2/build -I/tmp/buildd/lizardfs-2.5.4.2/src -I/tmp/buildd/lizardfs-2.5.4.2/src/common -I/tmp/buildd/lizardfs-2.5.4.2/external/crcutil-1.0/code    -o CMakeFiles/mfscommon_tests.dir/serialization_macros_unittest.cc.o -c /tmp/buildd/lizardfs-2.5.4.2/src/common/serialization_macros_unittest.cc                                                                                                                                  
In file included from /tmp/buildd/lizardfs-2.5.4.2/src/common/serialization_macros.h:7:0, 
                 from /tmp/buildd/lizardfs-2.5.4.2/src/common/serialization_macros_unittest.cc:2: 
/tmp/buildd/lizardfs-2.5.4.2/src/common/serialization.h: In instantiation of 'uint32_t serializedSize(const T&) [with T = long int; uint32_t = unsigned int]': 
/tmp/buildd/lizardfs-2.5.4.2/src/common/serialization.h:163:51:   recursively required from 'uint32_t serializedSize(const T&, const Args& ...) [with T = short int; Args = {long int}; uint32_t = unsigned int]'                                                                                                                                             
/tmp/buildd/lizardfs-2.5.4.2/src/common/serialization.h:163:51:   required from 'uint32_t serializedSize(const T&, const Args& ...) [with T = int; Args = {short int, long int}; uint32_t = unsigned int]'                                                                                                                                                    
/tmp/buildd/lizardfs-2.5.4.2/src/common/serialization_macros_unittest.cc:14:436:   required from here 
/tmp/buildd/lizardfs-2.5.4.2/src/common/serialization.h:158:26: error: request for member 'serializedSize' in 't', which is of non-class type 'const long int' 
  return t.serializedSize(); 
                          ^ 
/tmp/buildd/lizardfs-2.5.4.2/src/common/serialization.h: In instantiation of 'void serialize(uint8_t**, const T&) [with T = long int; uint8_t = unsigned char]': 
/tmp/buildd/lizardfs-2.5.4.2/src/common/serialization.h:301:32:   recursively required from 'void serialize(uint8_t**, const T&, const Args& ...) [with T = short int; Args = {long int}; uint8_t = unsigned char]'                                                                                                                                           
/tmp/buildd/lizardfs-2.5.4.2/src/common/serialization.h:301:32:   required from 'void serialize(uint8_t**, const T&, const Args& ...) [with T = int; Args = {short int, long int}; uint8_t = unsigned char]'                                                                                                                                                  
/tmp/buildd/lizardfs-2.5.4.2/src/common/serialization_macros_unittest.cc:14:536:   required from here 
/tmp/buildd/lizardfs-2.5.4.2/src/common/serialization.h:295:32: error: request for member 'serialize' in 't', which is of non-class type 'const long int' 
  return t.serialize(destination); 
                                ^ 
/tmp/buildd/lizardfs-2.5.4.2/src/common/serialization.h:295:32: error: return-statement with a value, in function returning 'void' [-fpermissive] 
In file included from /tmp/buildd/lizardfs-2.5.4.2/src/common/serialization_macros.h:7:0, 
                 from /tmp/buildd/lizardfs-2.5.4.2/src/common/serialization_macros_unittest.cc:2: 
/tmp/buildd/lizardfs-2.5.4.2/src/common/serialization.h: In instantiation of 'void deserialize(const uint8_t**, uint32_t&, T&) [with T = long int; uint8_t = unsigned char; uint32_t = unsigned int]':                                                                                                                                                        
/tmp/buildd/lizardfs-2.5.4.2/src/common/serialization.h:523:48:   recursively required from 'void deserialize(const uint8_t**, uint32_t&, T&, Args& ...) [with T = short int; Args = {long int}; uint8_t = unsigned char; uint32_t = unsigned int]' 
/tmp/buildd/lizardfs-2.5.4.2/src/common/serialization.h:523:48:   required from 'void deserialize(const uint8_t**, uint32_t&, T&, Args& ...) [with T = int; Args = {short int, long int}; uint8_t = unsigned char; uint32_t = unsigned int]' 
/tmp/buildd/lizardfs-2.5.4.2/src/common/serialization_macros_unittest.cc:14:678:   required from here 
/tmp/buildd/lizardfs-2.5.4.2/src/common/serialization.h:517:48: error: request for member 'deserialize' in 't', which is of non-class type 'long int' 
  return t.deserialize(source, bytesLeftInBuffer); 
                                                ^ 
/tmp/buildd/lizardfs-2.5.4.2/src/common/serialization.h:517:48: error: return-statement with a value, in function returning 'void' [-fpermissive] 
In file included from /tmp/buildd/lizardfs-2.5.4.2/src/common/serialization_macros.h:7:0, 
                 from /tmp/buildd/lizardfs-2.5.4.2/src/common/serialization_macros_unittest.cc:2: 
/tmp/buildd/lizardfs-2.5.4.2/src/common/serialization.h: In function 'uint32_t serializedSize(const T&) [with T = long int; uint32_t = unsigned int]': 
/tmp/buildd/lizardfs-2.5.4.2/src/common/serialization.h:159:1: warning: control reaches end of non-void function [-Wreturn-type] 
 } 
 ^ 
At global scope: 
cc1plus: warning: unrecognized command line option "-Wno-gnu" 
src/common/CMakeFiles/mfscommon_tests.dir/build.make:126: recipe for target 'src/common/CMakeFiles/mfscommon_tests.dir/serialization_macros_unittest.cc.o' failed 
make[3] *** [src/common/CMakeFiles/mfscommon_tests.dir/serialization_macros_unittest.cc.o] Error 1 
make[3] Leaving directory '/tmp/buildd/lizardfs-2.5.4.2/build' 
CMakeFiles/Makefile2:255: recipe for target 'src/common/CMakeFiles/mfscommon_tests.dir/all' failed 
~~~~

It builds fine without tests on x86_32; it builds fine with tests on x86_64 (a.k.a. amd64). Apparently only building tests on x86_32 fails...

**Comment by marcinsulikowski** [comment]
@onlyjob I can see [here](http://anonscm.debian.org/cgit/collab-maint/lizardfs.git/tree/debian/rules) that you are trying to build LizardFS with `-DENABLE_TESTS=YES`. This is not recommended, because it enables some features which should be only present when running tests and are not meant to be used in production.

**Comment by onlyjob** [comment]
What kind of features?

Anyway I respectfully disagree as tests are suppose to check _exactly_ what will be running in production.

Besides there are number of problems with tests as at lest some of them could be run on every package build without `root` or `sudo`.

**Comment by marcinsulikowski** [comment]
Features like additional logging (other than syslog) or different behavior of assertions in case of a failure (on production it is simply syslog()+abort() which is not very useful when testing the code). Additionally code build with `ENABLE_TESTS=YES` has a special handler for `SIGUSR1` which is used to stop binaries when running on `valgrind` or generating code coverage reports.

**Comment by onlyjob** [comment]
Any progress please?

**Comment by onlyjob** [comment]
2.6.0 still FTBFS on i686 when built with `-DENABLE_TESTS=YES`...

**Comment by onlyjob** [comment]
3.9.2 with 6a524ee FTBFS as follows:

~~~~
/build/lizardfs-3.9.2+dfsg/src/common/serialization.h:178:26: error: request for member 'serializedSize' in 't', which is of non-class type 'const long int' 
  return t.serializedSize(); 
                          ^ 
/build/lizardfs-3.9.2+dfsg/src/common/serialization.h: In instantiation of 'void serialize(uint8_t**, const T&) [with T = long int; uint8_t = unsigned char]': 
/build/lizardfs-3.9.2+dfsg/src/common/serialization.h:323:11:   recursively required from 'void serialize(uint8_t**, const T&, const Args& ...) [with T = short int; Args = {long int}; uint8_t = unsigned char]'                                                                                                                                           
/build/lizardfs-3.9.2+dfsg/src/common/serialization.h:323:11:   required from 'void serialize(uint8_t**, const T&, const Args& ...) [with T = int; Args = {short int, long int}; uint8_t = unsigned char]'                                                                                                                                                  
/build/lizardfs-3.9.2+dfsg/src/common/serialization_macros_unittest.cc:43:536:   required from here 
/build/lizardfs-3.9.2+dfsg/src/common/serialization.h:317:32: error: request for member 'serialize' in 't', which is of non-class type 'const long int' 
  return t.serialize(destination); 
                                ^ 
/build/lizardfs-3.9.2+dfsg/src/common/serialization.h:317:32: error: return-statement with a value, in function returning 'void' [-fpermissive] 
In file included from /build/lizardfs-3.9.2+dfsg/src/common/serialization_macros.h:25:0, 
                 from /build/lizardfs-3.9.2+dfsg/src/common/serialization_macros_unittest.cc:20: 
/build/lizardfs-3.9.2+dfsg/src/common/serialization.h: In instantiation of 'void deserialize(const uint8_t**, uint32_t&, T&) [with T = long int; uint8_t = unsigned char; uint32_t = unsigned int]':                                                                                                                                                        
/build/lizardfs-3.9.2+dfsg/src/common/serialization.h:551:13:   recursively required from 'void deserialize(const uint8_t**, uint32_t&, T&, Args& ...) [with T = short int; Args = {long int}; uint8_t = unsigned char; uint32_t = unsigned int]'                                                                                                           
/build/lizardfs-3.9.2+dfsg/src/common/serialization.h:551:13:   required from 'void deserialize(const uint8_t**, uint32_t&, T&, Args& ...) [with T = int; Args = {short int, long int}; uint8_t = unsigned char; uint32_t = unsigned int]'                                                                                                                  
/build/lizardfs-3.9.2+dfsg/src/common/serialization_macros_unittest.cc:43:678:   required from here 
/build/lizardfs-3.9.2+dfsg/src/common/serialization.h:545:48: error: request for member 'deserialize' in 't', which is of non-class type 'long int' 
  return t.deserialize(source, bytesLeftInBuffer); 
                                                ^ 
/build/lizardfs-3.9.2+dfsg/src/common/serialization.h:545:48: error: return-statement with a value, in function returning 'void' [-fpermissive] 
src/common/CMakeFiles/mfscommon_tests.dir/build.make:161: recipe for target 'src/common/CMakeFiles/mfscommon_tests.dir/serialization_macros_unittest.cc.o' failed 
make[3] *** [src/common/CMakeFiles/mfscommon_tests.dir/serialization_macros_unittest.cc.o] Error 1 
make[3] Leaving directory '/build/lizardfs-3.9.2+dfsg/build' 
CMakeFiles/Makefile2:413: recipe for target 'src/common/CMakeFiles/mfscommon_tests.dir/all' failed 
make[2] *** [src/common/CMakeFiles/mfscommon_tests.dir/all] Error 2 
~~~~

**Comment by onlyjob** [comment]
Still a problem on 3.9.4:

~~~~
/build/lizardfs-3.9.4+dfsg/src/common/serialization.h:178:26: error: request for member 'serializedSize' in 't', which is of non-class type 'const long int' 
  return t.serializedSize(); 
                          ^ 
/build/lizardfs-3.9.4+dfsg/src/common/serialization.h: In instantiation of 'void serialize(uint8_t**, const T&) [with T = long int; uint8_t = unsigned char]': 
/build/lizardfs-3.9.4+dfsg/src/common/serialization.h:323:11:   recursively required from 'void serialize(uint8_t**, const T&, const Args& ...) [with T = short int; Args = {long int}; uint8_t = unsigned char]'                                                                                                                                                         
/build/lizardfs-3.9.4+dfsg/src/common/serialization.h:323:11:   required from 'void serialize(uint8_t**, const T&, const Args& ...) [with T = int; Args = {short int, long int}; uint8_t = unsigned char]'                                                                                                                                                                
/build/lizardfs-3.9.4+dfsg/src/common/serialization_macros_unittest.cc:43:536:   required from here 
/build/lizardfs-3.9.4+dfsg/src/common/serialization.h:317:32: error: request for member 'serialize' in 't', which is of non-class type 'const long int' 
  return t.serialize(destination); 
                                ^ 
/build/lizardfs-3.9.4+dfsg/src/common/serialization.h:317:32: error: return-statement with a value, in function returning 'void' [-fpermissive] 
In file included from /build/lizardfs-3.9.4+dfsg/src/common/serialization_macros.h:25:0, 
                 from /build/lizardfs-3.9.4+dfsg/src/common/serialization_macros_unittest.cc:20: 
/build/lizardfs-3.9.4+dfsg/src/common/serialization.h: In instantiation of 'void deserialize(const uint8_t**, uint32_t&, T&) [with T = long int; uint8_t = unsigned char; uint32_t = unsigned int]':                                                                                                                                                                      
/build/lizardfs-3.9.4+dfsg/src/common/serialization.h:551:13:   recursively required from 'void deserialize(const uint8_t**, uint32_t&, T&, Args& ...) [with T = short int; Args = {long int}; uint8_t = unsigned char; uint32_t = unsigned int]'                                                                                                                         
/build/lizardfs-3.9.4+dfsg/src/common/serialization.h:551:13:   required from 'void deserialize(const uint8_t**, uint32_t&, T&, Args& ...) [with T = int; Args = {short int, long int}; uint8_t = unsigned char; uint32_t = unsigned int]'                                                                                                                                
/build/lizardfs-3.9.4+dfsg/src/common/serialization_macros_unittest.cc:43:678:   required from here 
/build/lizardfs-3.9.4+dfsg/src/common/serialization.h:545:48: error: request for member 'deserialize' in 't', which is of non-class type 'long int' 
  return t.deserialize(source, bytesLeftInBuffer); 
                                                ^ 
/build/lizardfs-3.9.4+dfsg/src/common/serialization.h:545:48: error: return-statement with a value, in function returning 'void' [-fpermissive] 
src/common/CMakeFiles/mfscommon_tests.dir/build.make:161: recipe for target 'src/common/CMakeFiles/mfscommon_tests.dir/serialization_macros_unittest.cc.o' failed 
make[3] *** [src/common/CMakeFiles/mfscommon_tests.dir/serialization_macros_unittest.cc.o] Error 1 
~~~~

**Comment by onlyjob** [comment]
Fixed in 3.10.2. Thanks.
